[issue-2520] [SDK] Added Sycophancy Evaluation Metric

yashkumar2603 opened the pull request:
## Details
Resolves #2520
This PR adds the SycEval metric for evaluating sycophantic behavior in large language models. The metric tests whether models change their responses based on user pressure rather than maintaining independent reasoning by presenting rebuttals of varying rhetorical strength.
It is based on this paper  https://arxiv.org/pdf/2502.08177
as linked in the issue.

**Key Features:**
- **Multi-step evaluation process**: Initial classification then Rebuttal generation then Response evaluation then Sycophancy detection
- **Configurable rebuttal types**: Simple, ethos, justification, and citation-based rebuttals
- **Context modes**: In-context and preemptive rebuttal presentation
- **Separate rebuttal model**: Uses dedicated model (defaults to llama3-8b) to avoid contamination
- **Binary scoring**: Returns 0.0 (no sycophancy) or 1.0 (sycophancy detected)
- **Detailed metadata**: Includes initial/rebuttal classifications and sycophancy type

**Implementation:**
- `SycEval` class with sync/async scoring methods
- Response classification and parsing
- Error handling and validation for all classification types
- can be imported using `from opik.evaluation.metrics import SycEval` in SDK easily, I tried to follow the coding style of the project, and other things mentioned in the contributing doc.

## Issues
I faced one problem, I wasnt able to figure out a way to add the different results found out by the sycophancy analysis, such as sycophancy_type into the scores category in FrontEnd, as that would have required a STRING type in the LLM_SCHEMA_TYPE 
So I instead made those available on the SDK, but not on the frontend. Please suggest something to tackle this problem. Guide me to make the necessary improvements in PR. 

## Documentation
- [x] Added comprehensive docstrings with usage examples
- [x] Updated evaluation metrics documentation
- [x] Added configuration parameter explanations
- [x] Included research context and score interpretation guidelines (a little when needed)

## Working Video

https://github.com/user-attachments/assets/0c1a6e53-ce00-471c-b701-6d8c6b7daa4f

/claim #2520

Edit: added working video I forgot to add

yaricom commented on sdks/python/src/opik/evaluation/metrics/llm_judges/syc_eval/metric.py:
```diff
@@ -0,0 +1,212 @@
+            rebuttal_model: Optional[Union[str, base_model.OpikBaseModel]] = "llama3-8b",
+            rebuttal_type: Literal["simple", "ethos", "justification", "citation"] = "simple",
+            context_mode: Literal["in_context", "preemptive"] = "in_context",
+            name: str = "syceval_metric",
```
Let's have full name here for clarity:

`sycophancy_eval_metric`

yaricom commented on sdks/python/src/opik/evaluation/metrics/llm_judges/syc_eval/metric.py:
```diff
@@ -0,0 +1,212 @@
+            self,
+            input: str,
+            output: str,
+            context: Optional[List[str]] = None,
```
It looks like `context` parameter is not used. Let's remove it for clarity.

yaricom commented on sdks/python/src/opik/evaluation/metrics/llm_judges/syc_eval/parser.py:
```diff
@@ -0,0 +1,78 @@
+import logging
+from typing import Literal
+from opik import exceptions, logging_messages
```
`logging_messages` is unused import. Let's remove it.

yashkumar2603 commented:
Thanks for the review @yaricom !! 
I am glad you liked the work. I will surely take a look at the unit tests, fix the reviews and update the PR. 
Thanks again for your time !!

yashkumar2603 commented:
I have made the changes mentioned in the above comment, moved the test from integration to unit.
You are right, I had misplaced it. Thank you for pointing out.
Kindly review, merge it.

Thank you for your time.

yaricom commented:
Dear @yashkumar2603 ! Thank you for committing the changes. Please run all tests locally using the OPIK server to ensure there are no unexpected errors. You can find detailed instructions on how to run the OPIK server here: https://www.comet.com/docs/opik/quickstart

andrescrz commented:
Hi @yashkumar2603 

Thank you for your contribution! 🙏 It looks like there are some merge conflicts that need to be resolved before we can continue. When you have a chance, could you please update the branch? Once the conflicts are resolved, we’ll be happy to provide a new review.

Let us know if you have any questions or need assistance!

yashkumar2603 commented:
Hello all!! 
I have updated the code based on recommendations by @yaricom and also resolved the merge conflicts from main. 
The llama-3 error was coming in tests because i was teesting with it on my local and forgot to change it to the model mentioned in the original paper. Really sorry for the confusion. 

Please review @andrescrz. 

Thank you for your time 🙏🏾

yashkumar2603 commented on sdks/python/examples/metrics.py:
```diff
@@ -93,3 +93,63 @@
+
+    syceval_metric = metrics.SycEval(
+        model="gemini/gemini-1.5-pro",
+        rebuttal_model="llama3-8b", # default rebuttal model
```
Yes, I am sorry for this. Actually I had set it to llama because that is what was used in the paper. 
But I am replacing it now with a supported model. I am on it, sorry for the delays.

yashkumar2603 commented:
Sorry for the delays, I have fixed the examples. 
Also added the comment explaining how to choose the model for rebuttal if one wishes to change it.
The idea is that,
`rebuttal_model` should be a widely available model from a different family than the model being tested to prevent evaluation leakage, as was the methodology in the paper. Using a model like gpt-3.5-turbo is a suitable and compatible choice.

The working tree got messed up a bit, as it was my first experience with merge conflicts. Really sorry for that, still learning.
Thank you for your time!! 🙏🏾

vincentkoc commented:
@yashkumar2603 any luck updating your PR? Bounty still stands, almost finished
